### PR TITLE
feat(tianmu): support alter column length and column type (#411)

### LINKED ISSUE #411: feature: alter table need to support  alter column length and column type
**Is your feature request related to a problem? Please describe.**

i can not alter colum length and colun type
return ”ERROR 2233 (HY000): Be disgraceful to storage engine, operating is forbidden!“

**Describe the solution you'd like**

 need to support  alter column length and column type

**Describe alternatives you've considered**

use copy algorithm to support it

**Additional context**

### PR DESCRIPTION
**Pull request by DandreChen**
## Summary about this PR

Issue Number: close #411 

## Tests Check List

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [ ] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/handler/tianmu_handler.cpp [hunk 1]
```diff
@@ -1476,6 +1476,12 @@ enum_alter_inplace_result TianmuHandler::check_if_supported_inplace_alter([[mayb
                                                                            Alter_inplace_info *ha_alter_info) {
   if ((ha_alter_info->handler_flags & ~TIANMU_SUPPORTED_ALTER_ADD_DROP_ORDER) &&
       (ha_alter_info->handler_flags != TIANMU_SUPPORTED_ALTER_COLUMN_NAME)) {
+    // add alter column operation
```
Comment (RingsC) [inline storage/tianmu/handler/tianmu_handler.cpp]: Pls, refine these lines. ugly code.  
And test cases are not sufficient, refer to: 
https://dev.mysql.com/doc/internals/en/test-faults-macros.html

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/handler/tianmu_handler.cpp (hunk 1 above)
Comment (DandreChen) [reply storage/tianmu/handler/tianmu_handler.cpp]: update comment and add macros

**Review by isredstar — APPROVED** [review]

File: mysql-test/suite/tianmu/t/alter_table.test [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/alter_table.test:3]: Suggestion: disable the warnings or query log.

File: mysql-test/suite/tianmu/t/alter_table.test [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+CREATE TABLE `t_test` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `first_name` varchar(10),
+  `last_name` varchar(10),
+  `sex` varchar(5),
+  `score` int(11),
+  `copy_id` int(11),
+  PRIMARY KEY (`id`)
+) ENGINE=tianmu DEFAULT CHARSET=utf8;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/alter_table.test]: Checking the engine. 
Pls. refer to : https://www.cnblogs.com/sms-xc/p/3651901.html

File: mysql-test/suite/tianmu/t/alter_table.test [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+show create table t_test;
+
+#
+#CLOSE STRICT MODE
+#
+
+set sql_mode="NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION";
+
+alter table t_test modify first_name char(5);
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/alter_table.test:50]: how about alter to an exceeded length.

File: storage/tianmu/handler/tianmu_handler.cpp [hunk 5]
```diff
@@ -1474,11 +1474,18 @@ int TianmuHandler::reset() {
 
 enum_alter_inplace_result TianmuHandler::check_if_supported_inplace_alter([[maybe_unused]] TABLE *altered_table,
                                                                            Alter_inplace_info *ha_alter_info) {
+  DBUG_ENTER(__PRETTY_FUNCTION__);
   if ((ha_alter_info->handler_flags & ~TIANMU_SUPPORTED_ALTER_ADD_DROP_ORDER) &&
       (ha_alter_info->handler_flags != TIANMU_SUPPORTED_ALTER_COLUMN_NAME)) {
```
Comment (RingsC) [inline storage/tianmu/handler/tianmu_handler.cpp:1485]: return the code flow as soon as possible.

if (xxx) --not satisfied
   return xxxx;
if (yyy)
   return yyyy

 ....

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by RingsC — APPROVED** [review]
LGTM

**Pull request merged**